File: src/main/java/co/aplicared/jvm/juego/arbúcies/Game.java
package co.aplicared.jvm.juego.arbúcies;

import co.aplicared.jvm.juego.arbúcies.control.Keyboard;
import co.aplicared.jvm.juego.arbúcies.control.Mouse;
import co.aplicared.jvm.juego.arbúcies.entity.mob.Player;
import co.aplicared.jvm.juego.arbúcies.graphics.Screen;
import co.aplicared.jvm.juego.arbúcies.level.Level;
import co.aplicared.jvm.juego.arbúcies.level.TileCoord;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferStrategy;
import java.awt.image.BufferedImage;
import java.awt.image.DataBufferInt;

@Deprecated
public class Game extends Canvas implements Runnable {
    private static final long serialVersionUID = 1L;

    public static final String title = "Arbúcies";
    public static int width = 300, height = width / 16 * 10, scale = 3;
    private Thread _gameThread;
    private JFrame _gameFrame;

    private boolean _isRunning = false;

    private BufferedImage _image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
    private int[] _pixels = ((DataBufferInt) _image.getRaster().getDataBuffer()).getData();

    private Keyboard _key;
    private Level _level;
    private Screen _screen;
    private Player _player;
    private Mouse _mouse;

    public Game() {
        Dimension size = new Dimension(width * scale, height * scale);
        TileCoord playerSpawn = new TileCoord(20, 65);

        _gameFrame = new JFrame();
        setPreferredSize(size);

        _key = new Keyboard();
        _mouse = new Mouse();
        _level = Level.spawn;
        _screen = new Screen(width, height);
        _player = new Player(playerSpawn.getX(), playerSpawn.getY(), _key);
        _player.init(_level);

        addKeyListener(_key);
        addMouseListener(_mouse);
        addMouseMotionListener(_mouse);
    }

    public static void main(String[] args) {
        Game game = new Game();
        game._gameFrame.setResizable(false);
        game._gameFrame.setTitle(title);
        game._gameFrame.add(game);
        game._gameFrame.pack();
        game._gameFrame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        game._gameFrame.setLocationRelativeTo(null);
        game._gameFrame.setVisible(true);
        game._gameFrame.requestFocus();

        game.start();
    }

    @Override
    public void run() {
        if (System.getProperty("os.name").toLowerCase().equals("mac os x")) {
            System.out.println("macOS");
            System.setProperty("apple.laf.useScreenMenuBar", "true");
            System.setProperty("com.apple.mrj.application.apple.menu.about.name", title);
        }
        long lastTime = System.nanoTime();
        long timer = System.currentTimeMillis();
        double ns = 1000000000.0 / 60.0;
        double delta = 0.0;
        int frames = 0;
        int ticks = 0;
        requestFocus();

        while (_isRunning) {
            long now = System.nanoTime();
            delta += (now - lastTime) / ns;
            lastTime = now;

            while (delta >= 1) {
                tick();
                ticks++;
                delta--;
            }

            render();

            frames++;
            if (System.currentTimeMillis() - timer > 1000) {
                timer += 1000;
                _gameFrame.setTitle(title + " (J) || TPS: " + ticks + " | FPS: " + frames);
                ticks = 0;
                frames = 0;
            }
        }
        stop();
    }

    private void render() {
        BufferStrategy bs = getBufferStrategy();

        if (bs == null) {
            createBufferStrategy(3);
            return;
        }

        _screen.clear();

        int xScroll = _player.x - _screen.width / 2;
        int yScroll = _player.y - _screen.height / 2;

        _level.render(xScroll, yScroll, _screen);
        _player.render(_screen);

        for (int i = 0; i < _pixels.length; i++) {
            _pixels[i] = _screen.pixels[i];
        }

        Graphics g = bs.getDrawGraphics();
        g.drawImage(_image, 0, 0, getWidth(), getHeight(), null);
        g.setColor(Color.WHITE);
        g.setFont(new Font("Ubuntu", Font.PLAIN, 20));
        g.fillRect(Mouse.getMouseX() - 32, Mouse.getMouseY() - 32, 64, 64);
        g.drawString("Button: " + Mouse.getMouseB(), 80, 80);
        g.dispose();
        bs.show();
    }

    private void tick() {
        _key.update();
        _player.update();
    }

    public synchronized void start() {
        _isRunning = true;
        _gameThread = new Thread(this, "Arbúcies (MAIN_GAME)");
        _gameThread.start();
    }

    public synchronized void stop() {
        try {
            _gameThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/co/aplicared/jvm/juego/arbucies/Game.kt
package co.aplicared.jvm.juego.arbucies

import co.aplicared.jvm.juego.arbucies.control.Keyboard
import co.aplicared.jvm.juego.arbucies.graphics.Screen
import java.awt.Canvas
import java.awt.Dimension
import java.awt.image.BufferedImage
import java.awt.image.DataBufferInt
import javax.swing.JFrame

class Game : Runnable, Canvas() {

    companion object {
        val gWidth = 300
        val gHeight = gWidth / 16 * 10
        val scale = 3

        val title = "Arbúcies"

        @JvmStatic
        fun main(args: Array<String>) {
            val game = Game()
            game._frame.isResizable = false
            game._frame.title = title
            game._frame.add(game)
            game._frame.pack()
            game._frame.requestFocus()
            game._frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
            game._frame.setLocationRelativeTo(null)
            game._frame.isVisible = true

            game.start()
        }
    }

    var _isRunning = false

    private lateinit var _gameThread: Thread
    private var _frame: JFrame
    private var _screen: Screen

    private var _key: Keyboard

    private val _image = BufferedImage(gWidth, gHeight, BufferedImage.TYPE_INT_RGB)
    private val _pixels = (_image.raster.dataBuffer as DataBufferInt).data

    private var xPos = 0
    private var yPos = 0

    init {
        val size = Dimension(gWidth * scale, gHeight * scale)
        _screen = Screen(gWidth, gHeight)
        _frame = JFrame()
        preferredSize = size
        _key = Keyboard()

        addKeyListener(_key)
    }

    override fun run(): Unit {
        var lastTime = System.nanoTime()
        var timer = System.currentTimeMillis()
        val ns = 1000000000.0 / 60.0
        var delta = 0.0
        var frames = 0
        var ticks = 0

        while (_isRunning) {
            val now = System.nanoTime()
            delta += (now - lastTime) / ns
            lastTime = now

            while (delta >= 1) {
                tick()
                ticks++
                delta--
            }

            render()
            frames++
            if (System.currentTimeMillis() - timer > 1000) {
                timer += 1000
                _frame.title = "$title || TPS: $ticks | FPS: $frames"
                ticks = 0
                frames = 0
            }
        }
        stop()
    }

    @Synchronized
    fun start() {
        _gameThread = Thread(this, "Arbúcies Game Thread")
        _gameThread.start()
        _isRunning = true
    }

    @Synchronized
    fun stop() {
        try {
            _gameThread.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    fun tick() {
        _key.update()
    }

    fun render() {
        val bs = bufferStrategy
        if (bs == null) {
            createBufferStrategy(3)
            return
        }

        _screen.clear()
        _screen.render()

        System.arraycopy(_screen.pixels, 0, _pixels, 0, _pixels.size)

        val g = bs.drawGraphics
        g.drawImage(_image, 0, 0, width, height, null)
        g.dispose()
        bs.show()
    }
}